Introduce StructuredSerializeInternal forStorage parameter

youennf opened the pull request:
#### 664fe9493a72508a29a3e6a3018bd55d4fca8985
<pre>
Introduce StructuredSerializeInternal forStorage parameter
<a href="https://bugs.webkit.org/show_bug.cgi?id=246744">https://bugs.webkit.org/show_bug.cgi?id=246744</a>
rdar://problem/101367447

Reviewed by Alex Christensen.

We introduce a SerializationForStorage (No/Yes) parameter to SerializedScriptValue::create.
This allows to throw in case we are serializing objects that are serializable but are throwing for forStorage=yes cases.
Make use of SerializationForStorage parameter for WebCodecsEncodedVideoChunk as per <a href="https://w3c.github.io/webcodecs/#encodedvideochunk-serialization.">https://w3c.github.io/webcodecs/#encodedvideochunk-serialization.</a>

The default is SerializationForStorage = No.
We set SerializationForStorage = Yes for IDB and IDLSerializedScriptValue, which is used for History pushState.
Both IDB and History.pushState are using <a href="https://html.spec.whatwg.org/multipage/structured-data.html#structuredserializeforstorage.">https://html.spec.whatwg.org/multipage/structured-data.html#structuredserializeforstorage.</a>

Covered by added tests.

* LayoutTests/imported/w3c/web-platform-tests/webcodecs/chunk-serialization.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/chunk-serialization.any.js:
(test):
* Source/WebCore/Modules/indexeddb/IDBCursor.cpp:
(WebCore::IDBCursor::update):
* Source/WebCore/Modules/indexeddb/IDBObjectStore.cpp:
(WebCore::IDBObjectStore::putOrAdd):
* Source/WebCore/Modules/webaudio/AudioWorkletNode.cpp:
(WebCore::AudioWorkletNode::create):
* Source/WebCore/bindings/js/JSDOMConvertSerializedScriptValue.h:
(WebCore::Converter&lt;IDLSerializedScriptValue&lt;T&gt;&gt;::convert):
* Source/WebCore/bindings/js/JSDOMWrapper.cpp:
(WebCore::cloneAcrossWorlds):
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::serialize):
(WebCore::CloneSerializer::CloneSerializer):
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::SerializedScriptValue::create):
* Source/WebCore/bindings/js/SerializedScriptValue.h:
(WebCore::SerializedScriptValue::convert):
* Source/WebCore/dom/ErrorEvent.cpp:
(WebCore::ErrorEvent::trySerializeError):
* Source/WebCore/dom/PopStateEvent.cpp:
(WebCore::PopStateEvent::trySerializeState):
* Source/WebCore/page/DOMWindow.cpp:
(WebCore::DOMWindow::postMessage):
* Source/WebCore/page/WindowOrWorkerGlobalScope.cpp:
(WebCore::WindowOrWorkerGlobalScope::structuredClone):
* Source/WebCore/workers/DedicatedWorkerGlobalScope.cpp:
(WebCore::DedicatedWorkerGlobalScope::postMessage):
* Source/WebCore/workers/Worker.cpp:
(WebCore::Worker::postMessage):
* Source/WebCore/workers/service/ExtendableMessageEvent.cpp:
(WebCore::ExtendableMessageEvent::ExtendableMessageEvent):
* Source/WebCore/workers/service/ServiceWorker.cpp:
(WebCore::ServiceWorker::postMessage):
* Source/WebCore/workers/service/ServiceWorkerClient.cpp:
(WebCore::ServiceWorkerClient::postMessage):

Canonical link: <a href="https://commits.webkit.org/255788@main">https://commits.webkit.org/255788@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/58063337133a0ebc754f460d06d0fd65d7e456ae

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93590 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2783 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24228 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103242 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163562 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97586 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2796 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31067 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85941 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99315 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99252 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1984 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80033 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/29022 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83906 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/83653 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/71975 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37450 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/17503 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35289 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18762 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4002 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39164 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41281 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41102 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37993 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/bindings/js/SerializedScriptValue.h:
```diff
@@ -68,6 +68,7 @@ class WebCodecsEncodedVideoChunkStorage;
 
 enum class SerializationErrorMode { NonThrowing, Throwing };
 enum class SerializationContext { Default, WorkerPostMessage, WindowPostMessage };
+enum class SerializationForStorage { No, Yes };
```
: bool

youennf replied:
OK, will change

achristensen07 commented on LayoutTests/imported/w3c/web-platform-tests/webcodecs/chunk-serialization.any.js:
```diff
@@ -72,3 +72,9 @@ async_test(t => {
+test(() => {
+  const chunk = createDefaultChunk("video", defaultVideoInit);
+  if (window.history)
+      assert_throws_dom("DataCloneError", () => history.pushState({ chunk }, null));
```
The spec says a TypeError.

youennf replied:
We are changing this, see https://github.com/w3c/webcodecs/issues/589

youennf committed:
Introduce StructuredSerializeInternal forStorage parameter
https://bugs.webkit.org/show_bug.cgi?id=246744
rdar://problem/101367447

Reviewed by Alex Christensen.

We introduce a SerializationForStorage (No/Yes) parameter to SerializedScriptValue::create.
This allows to throw in case we are serializing objects that are serializable but are throwing for forStorage=yes cases.
Make use of SerializationForStorage parameter for WebCodecsEncodedVideoChunk as per https://w3c.github.io/webcodecs/#encodedvideochunk-serialization.

The default is SerializationForStorage = No.
We set SerializationForStorage = Yes for IDB and IDLSerializedScriptValue, which is used for History pushState.
Both IDB and History.pushState are using https://html.spec.whatwg.org/multipage/structured-data.html#structuredserializeforstorage.

Covered by added tests.

* LayoutTests/imported/w3c/web-platform-tests/webcodecs/chunk-serialization.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/chunk-serialization.any.js:
(test):
* Source/WebCore/Modules/indexeddb/IDBCursor.cpp:
(WebCore::IDBCursor::update):
* Source/WebCore/Modules/indexeddb/IDBObjectStore.cpp:
(WebCore::IDBObjectStore::putOrAdd):
* Source/WebCore/Modules/webaudio/AudioWorkletNode.cpp:
(WebCore::AudioWorkletNode::create):
* Source/WebCore/bindings/js/JSDOMConvertSerializedScriptValue.h:
(WebCore::Converter<IDLSerializedScriptValue<T>>::convert):
* Source/WebCore/bindings/js/JSDOMWrapper.cpp:
(WebCore::cloneAcrossWorlds):
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::serialize):
(WebCore::CloneSerializer::CloneSerializer):
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::SerializedScriptValue::create):
* Source/WebCore/bindings/js/SerializedScriptValue.h:
(WebCore::SerializedScriptValue::convert):
* Source/WebCore/dom/ErrorEvent.cpp:
(WebCore::ErrorEvent::trySerializeError):
* Source/WebCore/dom/PopStateEvent.cpp:
(WebCore::PopStateEvent::trySerializeState):
* Source/WebCore/page/DOMWindow.cpp:
(WebCore::DOMWindow::postMessage):
* Source/WebCore/page/WindowOrWorkerGlobalScope.cpp:
(WebCore::WindowOrWorkerGlobalScope::structuredClone):
* Source/WebCore/workers/DedicatedWorkerGlobalScope.cpp:
(WebCore::DedicatedWorkerGlobalScope::postMessage):
* Source/WebCore/workers/Worker.cpp:
(WebCore::Worker::postMessage):
* Source/WebCore/workers/service/ExtendableMessageEvent.cpp:
(WebCore::ExtendableMessageEvent::ExtendableMessageEvent):
* Source/WebCore/workers/service/ServiceWorker.cpp:
(WebCore::ServiceWorker::postMessage):
* Source/WebCore/workers/service/ServiceWorkerClient.cpp:
(WebCore::ServiceWorkerClient::postMessage):

Canonical link: https://commits.webkit.org/255788@main
